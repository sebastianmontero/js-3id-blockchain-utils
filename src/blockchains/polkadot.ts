import {BlockchainHandler, BlockchainHandlerOpts} from "../blockchain-handler";
import {AccountID} from "caip";
import { getConsentMessage, LinkProof} from "../utils";
import { signatureVerify } from '@polkadot/util-crypto'
import * as uint8arrays from 'uint8arrays'

const namespace = 'polkadot'

const stringHex = (str: string): string => `0x${uint8arrays.toString(uint8arrays.fromString(str), 'base16')}`

async function createLink (did: string, account: AccountID, signer: any, opts: BlockchainHandlerOpts): Promise<LinkProof> {
    const linkMessageHex = stringHex(getConsentMessage(did, false).message)
    const res = await signer.signRaw({address: account.address, data: linkMessageHex})
    const proof: LinkProof = {
        version: 2,
        type: 'eoa',
        message: linkMessageHex,
        signature: res.signature,
        account: account.toString()
    }
    return proof
}

async function authenticate(message: string, account: AccountID, signer: any): Promise<string> {
    const res = await signer.signRaw({address: account.address, data: stringHex(message)})
    return res.signature
}

async function validateLink (proof: LinkProof): Promise<LinkProof | null> {
    const address = new AccountID(proof.account).address
    const res = await signatureVerify(proof.message, proof.signature, address)
    return res.isValid ? proof : null
}

const Handler: BlockchainHandler = {
    namespace,
    authenticate,
    validateLink,
    createLink
}

export default Handler
