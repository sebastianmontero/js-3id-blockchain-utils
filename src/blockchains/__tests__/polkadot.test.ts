import polkadot from "../polkadot";
import { AccountID } from "caip";
import { Signer, SignerResult } from '@polkadot/api/types';
import { KeyringPair } from '@polkadot/keyring/types';
import { SignerPayloadRaw } from '@polkadot/types/types';
import { TypeRegistry } from '@polkadot/types/create';
import createTestKeyring from '@polkadot/keyring/testing';
import { assert, hexToU8a, u8aToHex } from '@polkadot/util';
import { waitReady } from '@polkadot/wasm-crypto';

const did = 'did:3:bafysdfwefwe'
const net = 'polkadot:b0a8d493285c2df73290dfb7e61f870f'
const seed = hexToU8a('0xabf8e00000000000000000000000000000000000000000000000000000000000')

// primary and supported by polkadot extension
const keyringSr25519 = createTestKeyring({ type: 'sr25519' })
const keyringEd25519 = createTestKeyring({ type: 'ed25519' })
const keyringSecp256k = createTestKeyring({ type: 'ecdsa' })

const addressToAccountID = (address: string): AccountID => new AccountID(`${address}@${net}`)

class SingleAccountSigner implements Signer {
    readonly #keyringPair: KeyringPair;

    constructor (registry: TypeRegistry, keyringPair: KeyringPair) {
        this.#keyringPair = keyringPair
    }

    public async signRaw ({ address, data }: SignerPayloadRaw): Promise<SignerResult> {
        assert(address === this.#keyringPair.address, 'Signer does not have the keyringPair');
    
        return new Promise((resolve): void => {
            const signature = u8aToHex(this.#keyringPair.sign(hexToU8a(data)));
            resolve({ id: 1, signature })
        })
    }
}

describe('Blockchain: Polkadot', () => {
    const registry = new TypeRegistry()
    let keyPairSr25519, keyPairEd25519, keyPairSecp256k 

    beforeAll(async () => {
        await waitReady();
        keyPairSr25519 = keyringSr25519.addFromSeed(seed)
        keyPairEd25519 = keyringEd25519.addFromSeed(seed)
        keyPairSecp256k = keyringSecp256k.addFromSeed(seed)
    })

    describe('createLink', () => {
        test('create proof with sr25519', async () => {
            const account = addressToAccountID(keyPairSr25519.address)
            const provider = new SingleAccountSigner(registry, keyPairSr25519)
            const proof = await polkadot.createLink(did, account, provider)
            console.log(proof)
            expect(proof).toMatchSnapshot()
        })
        test('create proof with ed25519', async () => {
            const account = addressToAccountID(keyPairEd25519.address)
            const provider = new SingleAccountSigner(registry, keyPairEd25519)
            const proof = await polkadot.createLink(did, account, provider)
            expect(proof).toMatchSnapshot()
        })
        test('create proof with secp256k', async () => {
            const account = addressToAccountID(keyPairSecp256k.address)
            const provider = new SingleAccountSigner(registry, keyPairSecp256k)
            const proof = await polkadot.createLink(did, account, provider)
            expect(proof).toMatchSnapshot()
        })
    })

    describe('validateLink', () => {
        test('validate proof with sr25519', async () => {
            const account = addressToAccountID(keyPairSr25519.address)
            const provider = new SingleAccountSigner(registry, keyPairSr25519)
            const proof = await polkadot.createLink(did, account, provider)
            await expect(polkadot.validateLink(proof)).resolves.toEqual(proof)
        })
        test('validate proof with ed25519', async () => {
            const account = addressToAccountID(keyPairEd25519.address)
            const provider = new SingleAccountSigner(registry, keyPairEd25519)
            const proof = await polkadot.createLink(did, account, provider)
            await expect(polkadot.validateLink(proof)).resolves.toEqual(proof)
        })
        test('validate proof with secp256k', async () => {
            const account = addressToAccountID(keyPairSecp256k.address)
            const provider = new SingleAccountSigner(registry, keyPairSecp256k)
            const proof = await polkadot.createLink(did, account, provider)
            await expect(polkadot.validateLink(proof)).resolves.toEqual(proof)
        })
    })

    describe('authenticate', () => {
        test('authenticate with sr25519', async () => {
            const account = addressToAccountID(keyPairSr25519.address)
            const provider = new SingleAccountSigner(registry, keyPairSr25519)
            const authSecret = await polkadot.authenticate('msg', account, provider)
            expect(authSecret).toMatchSnapshot()
        })
        test('authenticate with ed25519', async () => {
            const account = addressToAccountID(keyPairEd25519.address)
            const provider = new SingleAccountSigner(registry, keyPairEd25519)
            const authSecret = await polkadot.authenticate('msg', account, provider)
            expect(authSecret).toMatchSnapshot()
        })
        test('authenticate with secp256k', async () => {
            const account = addressToAccountID(keyPairSecp256k.address)
            const provider = new SingleAccountSigner(registry, keyPairSecp256k)
            const authSecret = await polkadot.authenticate('msg', account, provider)
            expect(authSecret).toMatchSnapshot()
        })
    })

})
