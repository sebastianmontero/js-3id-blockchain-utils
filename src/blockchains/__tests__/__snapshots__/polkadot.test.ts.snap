// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockchain: Polkadot authenticate authenticate with ed25519 1`] = `"0x0c4649442e6acde7d1b4d7640d4330f5a8770394093cf77eda29136a555471a6c4981c3f6e6e22747abc473d539588463bcaaf53f870389e9bed279db4fe5e00"`;

exports[`Blockchain: Polkadot authenticate authenticate with secp256k 1`] = `"0x153026f67d745e8ac03ae181ff5e34cf7c8c1f0898d9573824ae89ff4243efe792747234a4706cc0becf3cddfffdc04a6ee70b9de7a841444a133dc6ed13777800"`;

exports[`Blockchain: Polkadot authenticate authenticate with sr25519 1`] = `"0x96a1ed246c33777e63aa800ed73bc3a95659f1cc307e807c11db0833f0864b3eae96a5adb26e3d91c6bbc57105370eb5e57b44b40b38ebd8c8ebfeb073eacc84"`;

exports[`Blockchain: Polkadot createLink create proof with ed25519 1`] = `
Object {
  "account": "5EEvMbuaJQxxPorGtzY2qUYqM3akRrnD36bZD4PomfpLdrtx@polkadot:b0a8d493285c2df73290dfb7e61f870f",
  "message": "0x4372656174652061206e6577206163636f756e74206c696e6b20746f20796f7572206964656e746974792e0a0a6469643a333a626166797364667765667765",
  "signature": "0xf457de8941e928ab8dfe6ac3d6b1fdc3b736b839b83fcd4f5e4697d20c3ff3d5c79654fc36dc401e34c311593b577753077f18474182f576c0215461277a1e00",
  "type": "eoa",
  "version": 2,
}
`;

exports[`Blockchain: Polkadot createLink create proof with secp256k 1`] = `
Object {
  "account": "5GuAnwx9uKSQxKZB8moxoaX4ceehoCYu4XYMjJkKtm2QDgqC@polkadot:b0a8d493285c2df73290dfb7e61f870f",
  "message": "0x4372656174652061206e6577206163636f756e74206c696e6b20746f20796f7572206964656e746974792e0a0a6469643a333a626166797364667765667765",
  "signature": "0x78e60152aacd9ab05ce3ad951a640d69f1ed0e24f8b81b78f695b9b1e3f2c83d5446b76dd7cbe5e18db41ddee3ffaef2b100003b8263e93d1d2a4ffd37b9cef601",
  "type": "eoa",
  "version": 2,
}
`;

exports[`Blockchain: Polkadot createLink create proof with sr25519 1`] = `
Object {
  "account": "5CXU7u72Gz9Jqo6FdAwzSsxX5JhdRwMj2UZ4xrAEoo3AyySx@polkadot:b0a8d493285c2df73290dfb7e61f870f",
  "message": "0x4372656174652061206e6577206163636f756e74206c696e6b20746f20796f7572206964656e746974792e0a0a6469643a333a626166797364667765667765",
  "signature": "0x50953d38a7cdb7a8f04e0c5399b8a0496a7c714a71e657908dfcbc6e319e87476b04be188d7e1b90ee21ea4a23b6f7c3bc72db18cd37832281847b3dd625538f",
  "type": "eoa",
  "version": 2,
}
`;
